lib/serializer/CSerializer.h

```diff
@@ -80,10 +80,11 @@ class DLL_LINKAGE CSerializer
 	TTypeVecMap vectors; //entry must be a pointer to vector containing pointers to the objects of key type
 
 public:
-	bool smartVectorMembersSerialization;
-	bool sendStackInstanceByIds;
+	bool smartVectorMembersSerialization = false;
+	bool sendStackInstanceByIds = false;
 
-	CSerializer();
+	///This constructor is required for linking with server
+	CSerializer() = default;
 	~CSerializer();
```

Comment:
so what's our stance on declaring default ctor/dtor explicitly? In the PR I see a mix of approaches.

Reply:
@kambala-decapitator we cannot skip declaration here, because then it will fail to link.

Reply:
I don't think I understand... If default constructor is removed, compiler would synthesise one automatically. How could that lead to a link error? Or is there another ctor in the class?

Reply:
I think we should drop explicitly defaulted constructor/destructor if possible.

In this particular case it is probably not necessary?
Generally, in order for compiler to generate destructor compiler needs access to all class members (NOT just forward-declaration).
Because of this, if (for example) your class have shared_ptr<XXX> as member, XXX can be forward declared in header. However to emit destructor, compiler will need access to actual declaration of XXX. Meaning, destructor for such class must be defined in a place that has XXX declaration visible. Normally, cpp file is such place.

I think same applies for constructor, but never encounter this myself before.

Also, I won't be surprised if there are few other cases related to dll linkage.

Reply:
@kambala-decapitator @IvanSavenko I am rechecking this now, if link failure will not remain after clang and gcc upgrade, there can be a compiler bug or my build bug (somewhere I can forgot to do clean-build).

Reply:
After upgrade it seems to linking normally without a constructor after clean-build.
If CI passes - I think constructor here can safely be removed. Changes pushed.

Reply:
CI passed, so, constructor removal is fine.

Reply:
the PR still has a mix of approaches. Can default ctors/dtors be removed from everywhere?

Reply:
@kambala-decapitator dtors - certainly not, there is a places when dtor should be explicitly set to default in *.cpp file.

Reply:
And also - when we need to annotate than destructor is virtual. This is done non-consistent now, but should be more consistent. 

About ctors - I thought I removed all defaulted ctors.

Reply:
Yeah, looks to be mostly consistent now.
Searching for word 'default' in PR I see following scenario:
1) Constructor defined in-place as default: required, since class has 2nd constructor, so compiler will normally delete default (0-params) constructor.
2) Destructor defined as default in header: required, since this is virtual destructor of a base class
3) Destructor defined as default in source file: required for compiler to generate destructors for members.

I see only one inconsistency: sometimes destructor is declared in header and sometimes in source. (case 2 & 3 from above). We *can* move all of them in .cpp, but not sure if this is better style-wise. Any opinions, @kambala-decapitator ?

Personally fine either way IMO.

Reply:
> And also - when we need to annotate than destructor is virtual

Well, those can be marked as `override`, (except for base class obviously). Already saw that in other PR from you. Looks OK I think.

Reply:
@IvanSavenko I mean your case 2) If I have not removed virtual or override destructor in derived class - it is my fault.

Reply:
declaring virtual dtor is clear.

> dtors - certainly not, there is a places when dtor should be explicitly set to default in *.cpp file.

I don't think I get which case is that (also case number 3)... Could you show an example? Here's an example with forward declaration which I thought was related to this, but it works fine either if you place default dtor in header or remove the dtor from code, in both cases the `die` message is printed:

```
// 1.h
struct A {
	~A();
};
```

```
// 1.cpp
#include "1.h"
#include <iostream>
A::~A() { std::cout << "die\n"; }
```

```
// 2.h
struct A;
struct B {
	A& a;
	
	B(A& a_);
	~B() = default;
};
```

```
#include "1.h"
#include "2.h"
B::B(A& a_) : a{a_} {}

int main()
{
	A a;
	B b{a};
	return 0;
}
```

Reply:
even if I move `main()` to another file and remove `#include "1.h"` from `2.cpp` there's still no compile/link error

Reply:
@kambala-decapitator to get a link error, you need to slightly adjust an example in file 2.h

```c++
// 2.h
struct A;
struct B {
	std::unique_ptr<A> a;
	
	B(A& a_);
	~B() = default;
};
```

Then it will be link error, because to destroy unique_ptr, you need full signature of A.

Reply:
now it's fully clear, thank you very much!